src/io/binary_parser.cpp

```diff
@@ -263,16 +266,43 @@ long long read_edges_inc(string filename, dd::FactorGraph &fg) {
     }
 
     // add variables to factors
-    if (fg.variables[variable_id].domain_type == DTYPE_BOOLEAN) {
-      fg.factors[factor_id].tmp_variables.push_back(dd::VariableInFactor(
-          variable_id, fg.variables[variable_id].upper_bound, variable_id,
-          position, ispositive));
-    } else {
-      fg.factors[factor_id].tmp_variables.push_back(dd::VariableInFactor(
-          variable_id, position, ispositive, equal_predicate));
-    }
+    fg.factors[factor_id].tmp_variables.push_back(dd::VariableInFactor(
+        variable_id, position, ispositive, equal_predicate));
     fg.variables[variable_id].tmp_factor_ids.push_back(factor_id);
   }
   file.close();
   return count;
 }
+
+// read domains for multinomial
+void read_domains(std::string filename, dd::FactorGraph &fg) {
+  ifstream file;
+  file.open(filename.c_str(), ios::in | ios::binary);
+  long id, value;
+  while (true) {
+    file.read((char *)&id, 8);
+    if (!file.good()) break;
+
+    id = be64toh(id);
+    dd::Variable &variable = fg.variables[id];
+
+    long domain_size;
+    file.read((char *)&domain_size, 8);
+    domain_size = be64toh(domain_size);
+    assert(variable.cardinality == domain_size);
+
+    variable.domain = new std::vector<int>(domain_size);
+    variable.domain_map = new std::unordered_map<int, int>();
+
+    for (int i = 0; i < domain_size; i++) {
+      file.read((char *)&value, 8);
+      value = be64toh(value);
+      (*variable.domain)[i] = value;
+    }
+
+    std::sort(variable.domain->begin(), variable.domain->end());
```

Comment:
If we get rid of `variable.domain` vector, this sorting will no longer be necessary here.